File: src/main/java/com/adinsight/agent/tools/AgentTool.java
package com.adinsight.agent.tools;

import com.adinsight.agent.session.AgentSession;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.concurrent.CompletableFuture;

public interface AgentTool<I> {
    ObjectMapper MAPPER = new ObjectMapper();

    String name();

    String description();

    Class<I> inputType();

    Map<String, Object> inputSchema();

    CompletableFuture<ToolResult> execute(I input, ToolContext context);

    default boolean isEnabled(AgentSession session) {
        return true;
    }

    default CompletableFuture<ToolPermission> checkPermission(I input, ToolContext context) {
        return CompletableFuture.completedFuture(ToolPermission.allow());
    }

    default CompletableFuture<Map<String, Object>> call(Map<String, Object> args, AgentSession session) {
        I input = MAPPER.convertValue(args == null ? Map.of() : args, inputType());
        ToolContext context = new ToolContext(session);
        return checkPermission(input, context)
                .thenCompose(permission -> {
                    if (permission.denied()) {
                        return CompletableFuture.completedFuture(ToolResult.failure(
                                permission.message() == null ? "permission denied" : permission.message()
                        ));
                    }
                    return execute(input, context);
                })
                .thenApply(this::shapeResult);
    }

    default Map<String, Object> toOpenAiToolDef() {
        Map<String, Object> function = new LinkedHashMap<>();
        function.put("name", name());
        function.put("description", description());
        function.put("parameters", inputSchema());

        Map<String, Object> definition = new LinkedHashMap<>();
        definition.put("type", "function");
        definition.put("function", function);
        return definition;
    }

    default Map<String, Object> shapeResult(ToolResult result) {
        Map<String, Object> shaped = new LinkedHashMap<>();
        shaped.put("ok", result.ok());
        shaped.put("data", result.data());
        shaped.put("metadata", result.metadata());
        if (result.error() != null) {
            shaped.put("error", result.error());
        }
        return shaped;
    }
}

File: src/tools/agent-tool.ts
import type { AgentSession } from "../session/agent-session.js";
import { allowTool, isToolDenied, type ToolPermission } from "./tool-permission.js";
import { toolContext, type ToolContext } from "./tool-context.js";
import type { ToolResult } from "./tool-result.js";

export type JsonSchema = Readonly<Record<string, unknown>>;
export type ToolArgs = Readonly<Record<string, unknown>>;

export type ShapedToolResult = Readonly<{
  ok: boolean;
  data?: unknown;
  error?: string;
  metadata: Readonly<Record<string, unknown>>;
}>;

export interface AgentTool<TInput = ToolArgs, TOutput = unknown> {
  readonly name: string;
  readonly description: string;
  readonly inputSchema: JsonSchema;

  parseInput(args: ToolArgs): TInput;

  execute(input: TInput, context: ToolContext): Promise<ToolResult<TOutput>>;

  isEnabled?(session: AgentSession): boolean;

  checkPermission?(input: TInput, context: ToolContext): Promise<ToolPermission>;
}

export async function callTool<TInput, TOutput>(
  tool: AgentTool<TInput, TOutput>,
  args: ToolArgs,
  session: AgentSession
): Promise<ShapedToolResult> {
  const input = tool.parseInput(args);
  const context = toolContext(session);
  const permission = tool.checkPermission == null
    ? allowTool()
    : await tool.checkPermission(input, context);

  if (isToolDenied(permission)) {
    return shapeToolResult({
      ok: false,
      error: permission.message,
      metadata: {}
    });
  }

  return shapeToolResult(await tool.execute(input, context));
}

export function shapeToolResult(result: ToolResult): ShapedToolResult {
  if (result.ok) {
    return {
      ok: true,
      data: result.data,
      metadata: result.metadata
    };
  }

  return {
    ok: false,
    error: result.error,
    metadata: result.metadata
  };
}
